File: problems-adapt/0001-0100/0009_quadruple-zero-sums/solution_meet_in_the_middle.java
import java.util.HashMap;
import java.util.Map;

class Solution {

    public int countQuadrupleZeroSums(int[] first, int[] second, int[] third, int[] fourth) {
        // Meet in the middle: a+b+c+d = 0 iff a+b = -(c+d), so index the
        // first two arrays' pair sums with multiplicities (not a set).
        Map<Integer, Integer> sums = new HashMap<>();
        for (int a : first) {
            for (int b : second) {
                sums.merge(a + b, 1, Integer::sum);
            }
        }
        int total = 0;
        // Each (c,d) pair adds the number of (a,b) pairs summing to its
        // negation; every zero tuple is counted once via its unique split.
        for (int c : third) {
            for (int d : fourth) {
                total += sums.getOrDefault(-(c + d), 0);
            }
        }
        return total;
    }
}

File: problems-adapt/0401-0500/0454_quadruple-zero-sums/solution_meet_in_the_middle.py
class Solution:
    def countQuadrupleZeroSums(self, first: list[int], second: list[int], third: list[int], fourth: list[int]) -> int:
        # Meet in the middle: a+b+c+d = 0 iff a+b = -(c+d), so index the
        # first two arrays' pair sums with multiplicities (not a set).
        sums = {}
        for a in first:
            for b in second:
                key = a + b
                sums[key] = sums.get(key, 0) + 1
        total = 0
        # Each (c,d) pair adds the number of (a,b) pairs summing to its
        # negation; every zero tuple is counted once via its unique split.
        for c in third:
            for d in fourth:
                total += sums.get(-(c + d), 0)
        return total
